keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1045 @@
+	Namespace          string `asn1:"utf8"`
+	ServiceAccountName string `asn1:"utf8"`
+	PodName            string `asn1:"utf8"`
+	PodUID             string `asn1:"utf8"`
+	NodeName           string `asn1:"utf8"`
+}
+```
+
+Certificates and certificate requests that embed this extension do not need to include any data in the standard X.509 name or subjectAlternateName fields. The CertificateSigningRequests generated by the kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.
+
+Utility functions for embedding and extracting this extension from `crypto/x509`
+Certificate and CertificateRequest objects are defined in
+`k8s.io/component-helpers/kubernetesx509`.  They are intended to be used both by
+core Kubernetes components and third-party software that needs to interact with
+these certificates.
+
+### PodIdentity Extension Node Restriction
+
+The noderestriction admission plugin will check the consistency of PodIdentity extensions in CertificateSigningRequest objects created by `system:node:` identities.  The creation request will only be admitted if the PodIdentity extension refers to a pod that is currently running on the node that requested the certificate.
+
+This upholds the Kubernetes node isolation guarantee by ensuring that
+compromising a single node only results in compromises of workloads that have
+actually been scheduled onto that node.
+
+<<[UNRESOLVED]>>
+Is it necessary or desirable to forbid non-kubelet identities from creating CertificateSigningRequest objects that embed a PodIdentity extension?
+<<[/UNRESOLVED]>>
+
+During Alpha, this capability will be behind the PodIdentityX509Support feature gate.
+
+### PodCertificate Projected Volume Sources
+
+The PodCertificate projected volume source instructs kubelet to inject a
+credential bundle consisting of a private key and certificate chain into the
+container fileystem.  kubelet handles generating the key, requesting a
+certificate from the named signer, and renewing the certificate before it
+expires.
+
+The kubelet supports only a few key generation strategies, each named by a string
+that encapsulates both the key type and any parameters necessary (for example,
+RSA modulus size).  The intention is to offer a a tasting menu of reasonable key
+choices, rather than offering a flexibility to a wide variety of parameters.
+Key types supported by kubelet are "RSA2048", "RSA3072", "RSA4096", "ECDSAP256",
+and "ECDSAP384".  If no key type is specified, kubelet defaults to "ECDSAP256".
+
+<<[UNRESOLVED]>>
+Should RSA2048 be offered?  @ahmedtd inclined to say no.  Are there any other key types that should be offered?
+<<[/UNRESOLVED]>>
+
+Once the key is generated, kubelet creates a CertificateSigningRequest that
+contains an X.509 certificate request with no data in the name or
+subjectAlternateName fields, but instead embeds an X.509 PodIdentity extension
+with the details of the pod that mounted the volume.  The signer is intended to
```

Comment:
See prev. comment about the OID PodIdentity extension - this section either needs some MUSTS or I'm not sure including the PodIdentity extension is much practical use.

- If the PodIdentity extension is a suggestion/hint to the issuer, not a rule - why define it at all as a x509 extension, if there's no guarantee the eventual cert will meaningfully relate to it/carry some/all of it at all?
- If all the fields in PodIdentity are optional (including `PodName`), what is a `named pod`, exactly?

Reply:
Re Point 1: The intention here is that the signer implementation will receive a fixed-format CSR from Kubelet, extract the pod identity (to the extent that it cares about it) from the CSR, and then issue the certificate it believes the pod should have.  For example, a hypothetical custom service mesh signer implementation would read the PodIdentity extension, but then issue a SPIFFE certificate that encodes the pod's identity in the way that it likes.

This division of responsibilities will keep us from having an explosion of configuration knobs on the certificate volume.

Re Point 2: In this section, it is Kubelet filling out the PodIdentity extension, so it will set all fields.  I'll clarify.

Reply:
>  For example, a hypothetical custom service mesh signer implementation would read the PodIdentity extension, but then issue a SPIFFE certificate that encodes the pod's identity in the way that it likes.

That makes sense. My concern was that if _all_ the fields in the PodIdentity are explicitly optional in the spec, it seems like the spec is encoding a situation where a custom implementation could not get enough information/hints from the provided PodIdentity to even sufficiently identify the pod, which seems not-ideal.

> Re Point 2: In this section, it is Kubelet filling out the PodIdentity extension, so it will set all fields. I'll clarify.

:+1:

Reply:
After API review, we decided to move to use a new PodCertificateRequest type.  This significantly cuts down the number of parties that need to care about the X.509 extension --- now it's basically only the built-in `kubernetes.io/kube-apiserver-client-pod` signer, and the kube-apiserver client certificate authenticator.

Random third-party signers will now just directly read pod details from top-level fields on the PodCertificateRequest object.